packages/stream-plugin/src/index.ts

```diff
@@ -0,0 +1,108 @@
+import {
+  createUniverseWatcher,
+  PluginActionCtx,
+  ProdoPlugin,
+} from "@prodo/core";
+
+const valueSymbol = Symbol("streamValues");
+
+interface StreamState {
+  unsubscribe: () => void;
+}
+
+export interface Stream<T> {
+  subscribe: (cb: (value: T) => void) => StreamState;
+}
+
+type UnStreams<T extends { [K in keyof T]?: Stream<any> }> = {
+  [K in keyof T]?: T[K] extends Stream<infer V> ? V : never;
+};
+
+// Stores the latest values
+export interface Universe<
+  T extends { [K in keyof T]: Stream<any> | undefined }
+> {
+  streams: UnStreams<T>;
+}
+
+interface State<T extends { [K in keyof T]: Stream<any> | undefined }> {
+  streams: T;
+  states: { [K in keyof T]?: StreamState };
+}
+
+export interface ViewCtx<T extends { [K in keyof T]?: Stream<any> }> {
+  streams: UnStreams<T>;
+}
+
+export interface ActionCtx<T extends { [K in keyof T]?: Stream<any> }> {
+  streams: T;
+  // Private to streamUpdate
+  [valueSymbol]: UnStreams<T>;
+}
+
+const init = <T extends { [K in keyof T]?: Stream<any> }>(
+  _config: {},
+  universe: Universe<T>,
+) => {
+  universe.streams = {};
+};
+
+const streamUpdate = <T extends { [K in keyof T]?: Stream<any> }>(
+  ctx: ActionCtx<T>,
+) => (key: keyof T, value: any) => {
+  ctx[valueSymbol][key] = value;
+};
+
+const prepareActionCtx = <T extends { [K in keyof T]?: Stream<any> }>(
+  state: State<T>,
+) => ({
+  ctx,
+  universe,
+}: {
+  ctx: PluginActionCtx<ActionCtx<T>> & ActionCtx<T>;
+  universe: Universe<T>;
+}) => {
+  ctx[valueSymbol] = universe.streams;
+  ctx.streams = new Proxy(state.streams, {
+    set: (obj, key: keyof T, value) => {
```

Comment:
We might want to allow reading the current value from an action.

Reply:
This (i.e. write a `Stream<T>`, read a `T`) is not possible to do with Typescript at the moment. See microsoft/TypeScript/issues/20846